src/parquet/file/metadata.h

```diff
@@ -0,0 +1,189 @@
+};
+
+class PARQUET_EXPORT FileMetaData {
+ public:
+  // API convenience to get a MetaData accessor
+  static std::unique_ptr<FileMetaData> GetFileMetaData(const uint8_t* metadata);
+
+  explicit FileMetaData(const uint8_t* metadata);
+  ~FileMetaData();
+
+  // file metadata
+  int64_t num_rows() const;
+  int num_row_groups() const;
+  int32_t version() const;
+  const std::string& created_by() const;
+  int num_schema_elements() const;
+  std::unique_ptr<RowGroupMetaData> GetRowGroupMetaData(int i) const;
+
+  // Return const-pointer to make it clear that this object is not to be copied
+  const SchemaDescriptor* schema() const;
+
+ private:
+  // PIMPL Idiom
+  class FileMetaDataImpl;
+  std::unique_ptr<FileMetaDataImpl> impl_;
+};
+
+// Builder API
+class PARQUET_EXPORT ColumnMetaDataBuilder {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<ColumnMetaDataBuilder> GetColumnMetaDataBuilder(
+      std::shared_ptr<WriterProperties> props, const ColumnDescriptor* column,
+      uint8_t* contents);
+
+  explicit ColumnMetaDataBuilder(std::shared_ptr<WriterProperties> props,
+      const ColumnDescriptor* column, uint8_t* contents);
+  ~ColumnMetaDataBuilder();
+
+  // column chunk
+  void set_file_path(const std::string& path);
+  // column metadata
+  void set_encodings(const std::set<Encoding::type>& val);
+  void SetStatistics(const ColumnStatistics& stats);
+
+  // commit the metadata
+  void Finish(int64_t num_values, int64_t dictonary_page_offset,
+      int64_t index_page_offset, int64_t data_page_offset, int64_t compressed_size,
+      int64_t uncompressed_size);
+
+ private:
+  // PIMPL Idiom
+  class ColumnMetaDataBuilderImpl;
+  std::unique_ptr<ColumnMetaDataBuilderImpl> impl_;
+};
+
+class PARQUET_EXPORT RowGroupMetaDataBuilder {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<RowGroupMetaDataBuilder> GetRowGroupMetaDataBuilder(
+      std::shared_ptr<WriterProperties> props, const SchemaDescriptor* schema_,
+      uint8_t* contents);
+
+  explicit RowGroupMetaDataBuilder(std::shared_ptr<WriterProperties> props,
+      const SchemaDescriptor* schema_, uint8_t* contents);
+  ~RowGroupMetaDataBuilder();
+
+  ColumnMetaDataBuilder* NextColumnMetaData();
+
+  // commit the metadata
+  void Finish(int64_t num_rows);
+
+ private:
+  // PIMPL Idiom
+  class RowGroupMetaDataBuilderImpl;
+  std::unique_ptr<RowGroupMetaDataBuilderImpl> impl_;
+};
+
+class PARQUET_EXPORT FileMetaDataBuilder {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<FileMetaDataBuilder> GetFileMetaDataBuilder(
+      const SchemaDescriptor* schema, std::shared_ptr<WriterProperties> props);
+
+  explicit FileMetaDataBuilder(
+      const SchemaDescriptor* schema, std::shared_ptr<WriterProperties> props);
+  ~FileMetaDataBuilder();
+
+  RowGroupMetaDataBuilder* AppendRowGroupMetaData();
+
+  // commit the metadata
+  void Finish();
+
+  const uint8_t* get_bytes();
```

Comment:
That's right. On closer examination I see that you have the `FileMetaData` struct tied to the lifetime of the `FileMetaDataBuilder`:

https://github.com/apache/parquet-cpp/pull/143/files#diff-1d365120bac4bbc47d68c1c93bec119eR478

Would it be better to have a `std::unique_ptr<format::FileMetaData>` inside so that ownership of that memory can be handed off to the (immutable) `parquet::FileMetaData` object when you call `Finish`?